### PR TITLE
use legacy restmapper against undiscoverable servers

### PR DESCRIPTION
**Pull request by deads2k**
Simple, coarse grained solution to https://bugzilla.redhat.com/show_bug.cgi?id=1360170#c4

@bparees inelegant, but easier than trying to fine tune fallback scenarios and it's easy to inspect.

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

File: pkg/cmd/util/clientcmd/factory.go [hunk 1]
```diff
@@ -512,6 +518,20 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
...
 
+// useDiscoveryRESTMapper checks the server version to see if its recent enough to have
+// enough discovery information avaiable to reliably build a RESTMapper.  If not, use the
+// hardcoded mapper in this client (legacy behavior)
+func useDiscoveryRESTMapper(serverVersion string) bool {
+	serverSemVer, err := semver.Parse(serverVersion[1:])
+	if err != nil {
+		return false
+	}
+	if serverSemVer.LT(semver.MustParse("1.3.0")) {
```
Comment (liggitt) [inline pkg/cmd/util/clientcmd/factory.go]: does this work for origin and ose and kubernetes?

File: pkg/cmd/util/clientcmd/factory.go [hunk 2]
```diff
@@ -512,6 +518,20 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 	return w
 }
 
+// useDiscoveryRESTMapper checks the server version to see if its recent enough to have
+// enough discovery information avaiable to reliably build a RESTMapper.  If not, use the
+// hardcoded mapper in this client (legacy behavior)
+func useDiscoveryRESTMapper(serverVersion string) bool {
+	serverSemVer, err := semver.Parse(serverVersion[1:])
```
Comment (bparees) [inline pkg/cmd/util/clientcmd/factory.go]: how does this handle emptystring (or whatever values ServerVersion() can return for legacy servers)?

File: pkg/cmd/util/clientcmd/factory.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > does this work for origin and ose and kubernetes?

Yes, it checks the kube version and we know which levels of kube we shipped with each level of origin.

File: pkg/cmd/util/clientcmd/factory.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > how does this handle emptystring (or whatever values ServerVersion() can return for legacy servers)?

We never shipped a level that gave back empty string.  If there was an error getting the version, we handle that before this call.

Comment (bparees) [inline pkg/cmd/util/clientcmd/factory.go]: ok

**Comment by deads2k** [comment]
@bparees any further comment?

**Comment by liggitt** [comment]
Does falling back to the legacy RestMapper mean we don't dynamically filter the `all` alias? Couldn't remember if those were connected

**Comment by deads2k** [comment]
> Does falling back to the legacy RestMapper mean we don't dynamically filter the all alias? Couldn't remember if those were connected

No, that's done by the shortcut expander layered on top.

**Comment by bparees** [comment]
@deads2k nope, lgtm.

**Comment by deads2k** [comment]
[merge]

**Pull request merged**